{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "d963a9ba-de16-441e-bcf1-e9f3fd636230",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "#from bioservices import Biomart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "639cf537-81d2-4c44-abf6-dd61cb7ab5bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Files have been loaded- count.txt, sample-annotation.txt and gene-annotation.txt\n"
     ]
    }
   ],
   "source": [
    "\n",
    "counts_df = pd.read_csv(\"counts.txt\", sep=\"\\t\")\n",
    "# converting counts_df to matrix\n",
    "counts_df.set_index(counts_df.columns[0], inplace=True)\n",
    "cts = counts_df.to_numpy()\n",
    "\n",
    "coldata = pd.read_csv(\"sample-annotation.txt\", sep=\"\\t\")\n",
    "coldata.set_index(coldata.columns[0], inplace=True)\n",
    "\n",
    "gene_anno = pd.read_csv(\"gene-annotation.txt\", sep=\"\\t\")\n",
    "gene_anno.set_index(gene_anno.columns[0], inplace=True)\n",
    "\n",
    "print(\"Files have been loaded- count.txt, sample-annotation.txt and gene-annotation.txt\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51ee7bda-1901-46ff-a1b8-bf420fd535e8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0ec499ce-fec3-4420-bbb2-b5db9ab292aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4385"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df.duplicated().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bfb52907-b241-4e19-9742-eb7646d4aa56",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total ensemble genes in gene-annotation.txt: 57992\n",
      "total non-duplicated ensemble genes in gene-annotation.txt: 57992\n",
      "Total genes annotated in gene-annotation.txt file: 25503\n",
      "Genes from counts.txt file that does not have annotation in gene-annotation.txt: 32489\n"
     ]
    }
   ],
   "source": [
    "# Missing data: gene-annotation.txt file is missing allot of genes which are presented in the counts.txt file\n",
    "print(\"total ensemble genes in gene-annotation.txt:\", cts.shape[0]) # 57992 genes\n",
    "print(\"total non-duplicated ensemble genes in gene-annotation.txt:\", len(np.unique(counts_df.index))) # 57992\n",
    "print(\"Total genes annotated in gene-annotation.txt file:\", gene_anno.shape[0]) # 25503 genes\n",
    "print(\"Genes from counts.txt file that does not have annotation in gene-annotation.txt:\", np.sum(~np.isin(counts_df.index, gene_anno.index))) # 32489 genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "a6960141-e3b8-4eed-923c-922be0fbc43e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import required modules\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "\n",
    "from biomart import BiomartServer\n",
    "# Connect to the Biomart database\n",
    "server = BiomartServer(\"http://www.ensembl.org/biomart\")\n",
    "db = server.databases['ENSEMBL_MART_ENSEMBL']\n",
    "dataset = db.datasets['hsapiens_gene_ensembl']\n",
    "\n",
    "\n",
    "ensembl_id_list = cts[:, 0].astype(str).tolist()\n",
    "ensembl_ids = \",\".join(ensembl_id_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "2443cd62-07ce-42a5-bbd4-2430dc83cc0d",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'BiomartDataset' object has no attribute 'query'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Input \u001b[1;32mIn [52]\u001b[0m, in \u001b[0;36m<cell line: 16>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     15\u001b[0m \u001b[38;5;66;03m# Loop over the batches and retrieve the annotations for each batch\u001b[39;00m\n\u001b[0;32m     16\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m batch \u001b[38;5;129;01min\u001b[39;00m batches:\n\u001b[0;32m     17\u001b[0m     \u001b[38;5;66;03m# Filter by Ensembl ID and retrieve the attributes\u001b[39;00m\n\u001b[1;32m---> 18\u001b[0m     query \u001b[38;5;241m=\u001b[39m \u001b[43mdataset\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mquery\u001b[49m(attributes\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mensembl_gene_id\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgene_biotype\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhgnc_symbol\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n\u001b[0;32m     19\u001b[0m     query\u001b[38;5;241m.\u001b[39madd_filter(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mensembl_gene_id\u001b[39m\u001b[38;5;124m\"\u001b[39m, batch, operator\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124min\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m     20\u001b[0m     annotLookup \u001b[38;5;241m=\u001b[39m query\u001b[38;5;241m.\u001b[39mresults()\n",
      "\u001b[1;31mAttributeError\u001b[0m: 'BiomartDataset' object has no attribute 'query'"
     ]
    }
   ],
   "source": [
    "# Connect to the Biomart database\n",
    "server = BiomartServer(\"http://www.ensembl.org/biomart\")\n",
    "db = server.databases['ENSEMBL_MART_ENSEMBL']\n",
    "dataset = db.datasets['hsapiens_gene_ensembl']\n",
    "\n",
    "# Ensemble genes (from count.txt) to annotate by Biomart\n",
    "ensLookup = cts[:, 0].astype(str).tolist()\n",
    "\n",
    "# Split the list of genes into batches of 1000 genes each\n",
    "batches = [ensLookup[i:i+1000] for i in range(0, len(ensLookup), 1000)]\n",
    "\n",
    "# Initialize a list to store the results\n",
    "annotLookup_list = []\n",
    "\n",
    "# Loop over the batches and retrieve the annotations for each batch\n",
    "for batch in batches:\n",
    "    # Filter by Ensembl ID and retrieve the attributes\n",
    "    query = dataset.query(attributes=[\"ensembl_gene_id\", \"gene_biotype\", \"hgnc_symbol\"])\n",
    "    query.add_filter(\"ensembl_gene_id\", batch, operator=\"in\")\n",
    "    annotLookup = query.results()\n",
    "    annotLookup_list.append(annotLookup)\n",
    "\n",
    "# Concatenate the results into a single dataframe\n",
    "annotLookup = pd.concat(annotLookup_list, axis=0)\n",
    "\n",
    "print(\"{} genes have been annotated by Biomart\".format(len(annotLookup)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "471d327e-5b28-44cd-8722-8d209d818052",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'numpy.ndarray' object has no attribute 'index'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Input \u001b[1;32mIn [53]\u001b[0m, in \u001b[0;36m<cell line: 10>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      7\u001b[0m dataset \u001b[38;5;241m=\u001b[39m db\u001b[38;5;241m.\u001b[39mdatasets[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhsapiens_gene_ensembl\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[0;32m      9\u001b[0m \u001b[38;5;66;03m# Ensemble genes (from count.txt) to annotate by Biomart\u001b[39;00m\n\u001b[1;32m---> 10\u001b[0m ensLookup \u001b[38;5;241m=\u001b[39m \u001b[43mcts\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mindex\u001b[49m\u001b[38;5;241m.\u001b[39mtolist()\n\u001b[0;32m     12\u001b[0m \u001b[38;5;66;03m# Split the list of genes into batches of 1000 genes each\u001b[39;00m\n\u001b[0;32m     13\u001b[0m batches \u001b[38;5;241m=\u001b[39m [ensLookup[i:i\u001b[38;5;241m+\u001b[39m\u001b[38;5;241m1000\u001b[39m] \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;241m0\u001b[39m, \u001b[38;5;28mlen\u001b[39m(ensLookup), \u001b[38;5;241m1000\u001b[39m)]\n",
      "\u001b[1;31mAttributeError\u001b[0m: 'numpy.ndarray' object has no attribute 'index'"
     ]
    }
   ],
   "source": [
    "from biomart import BiomartServer\n",
    "import pandas as pd\n",
    "\n",
    "# Connect to the Biomart database\n",
    "server = BiomartServer(\"http://www.ensembl.org/biomart\")\n",
    "db = server.databases['ENSEMBL_MART_ENSEMBL']\n",
    "dataset = db.datasets['hsapiens_gene_ensembl']\n",
    "\n",
    "# Ensemble genes (from count.txt) to annotate by Biomart\n",
    "ensLookup = cts.index.tolist()\n",
    "\n",
    "# Split the list of genes into batches of 1000 genes each\n",
    "batches = [ensLookup[i:i+1000] for i in range(0, len(ensLookup), 1000)]\n",
    "\n",
    "# Initialize a list to store the results\n",
    "annotLookup_list = []\n",
    "\n",
    "# Loop over the batches and retrieve the annotations for each batch\n",
    "for batch in batches:\n",
    "    # Filter by Ensembl ID and retrieve the attributes\n",
    "    query = dataset.query(attributes=[\"ensembl_gene_id\", \"gene_biotype\", \"hgnc_symbol\"])\n",
    "    query.add_filter(\"ensembl_gene_id\", batch, operator=\"in\")\n",
    "    annotLookup = query.results()\n",
    "    annotLookup_list.append(annotLookup)\n",
    "\n",
    "# Concatenate the results into a single dataframe\n",
    "annotLookup = pd.concat(annotLookup_list, axis=0)\n",
    "\n",
    "print(\"{} genes have been annotated by Biomart\".format(len(annotLookup)))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "30301a78-49de-4391-bb24-32eed3a205b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "\n",
    "# Connect to the Entrez database\n",
    "Entrez.email = \"libby3121@gmail.com\" # Set your email for Entrez\n",
    "\n",
    "ensembl_id_list = cts[:, 0].astype(str).tolist()\n",
    "ensembl_ids = \",\".join(ensembl_id_list)\n",
    "\n",
    "handle = Entrez.efetch(db=\"gene\", id=ensembl_ids, rettype=\"gb\", retmode=\"text\")\n",
    "gb_records = handle.read().split(\"//\\n\")[:-1]\n",
    "\n",
    "# Parse GenBank records and extract gene info\n",
    "annotLookup = []\n",
    "for gb_record in gb_records:\n",
    "    # Parse GenBank record\n",
    "    record = Entrez.read(Entrez.parse(gb_record))[0]\n",
    "\n",
    "    # Extract gene information\n",
    "    ensembl_gene_id = record[\"Dbxref\"].split(\":\")[1]\n",
    "    gene_biotype = record[\"gene_biotype\"]\n",
    "    hgnc_symbol = None\n",
    "    for xref in record[\"Entrezgene_xref\"]:\n",
    "        if xref[\"Dbname\"] == \"HGNC\":\n",
    "            hgnc_symbol = xref[\"Description\"]\n",
    "\n",
    "    # Add gene information to lookup table\n",
    "    annotLookup.append([ensembl_gene_id, gene_biotype, hgnc_symbol])\n",
    "\n",
    "# Convert lookup table to pandas DataFrame\n",
    "annotLookup = pd.DataFrame(annotLookup, columns=[\"ensembl_gene_id\", \"gene_biotype\", \"hgnc_symbol\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7e359b43-b834-458d-9b7a-135a5a6e17b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "handle = Entrez.efetch(db=\"gene\", id=ensembl_ids, rettype=\"gb\", retmode=\"text\")\n",
    "gb_records = handle.read().split(\"//\\n\")[:-1]\n",
    "\n",
    "print(len(gb_records))  # Check the number of records retrieved"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "59a5a540-931b-4958-9595-46755f014bac",
   "metadata": {},
   "outputs": [
    {
     "ename": "PackageNotInstalledError",
     "evalue": "The R package \"biomaRt\" is not installed.",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mPackageNotInstalledError\u001b[0m                  Traceback (most recent call last)",
      "Input \u001b[1;32mIn [47]\u001b[0m, in \u001b[0;36m<cell line: 3>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;66;03m# Load the biomaRt library\u001b[39;00m\n\u001b[0;32m      2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mrpy2\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mrobjects\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mpackages\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m importr\n\u001b[1;32m----> 3\u001b[0m biomaRt \u001b[38;5;241m=\u001b[39m \u001b[43mimportr\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mbiomaRt\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[0;32m      5\u001b[0m \u001b[38;5;66;03m# Connect to the biomart database\u001b[39;00m\n\u001b[0;32m      6\u001b[0m ensembl \u001b[38;5;241m=\u001b[39m biomaRt\u001b[38;5;241m.\u001b[39museMart(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mENSEMBL_MART_ENSEMBL\u001b[39m\u001b[38;5;124m\"\u001b[39m, dataset\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhsapiens_gene_ensembl\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[1;32m~\\OneDrive - virusight.co.il\\Documents\\synthetic\\lib\\site-packages\\rpy2\\robjects\\packages.py:472\u001b[0m, in \u001b[0;36mimportr\u001b[1;34m(name, lib_loc, robject_translations, signature_translation, suppress_messages, on_conflict, symbol_r2python, symbol_resolve, data)\u001b[0m\n\u001b[0;32m    440\u001b[0m \u001b[38;5;124;03m\"\"\" Import an R package.\u001b[39;00m\n\u001b[0;32m    441\u001b[0m \n\u001b[0;32m    442\u001b[0m \u001b[38;5;124;03mArguments:\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    468\u001b[0m \n\u001b[0;32m    469\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    471\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m isinstalled(name, lib_loc\u001b[38;5;241m=\u001b[39mlib_loc):\n\u001b[1;32m--> 472\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m PackageNotInstalledError(\n\u001b[0;32m    473\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mThe R package \u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m is not installed.\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m%\u001b[39m name\n\u001b[0;32m    474\u001b[0m     )\n\u001b[0;32m    476\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m suppress_messages:\n\u001b[0;32m    477\u001b[0m     ok \u001b[38;5;241m=\u001b[39m quiet_require(name, lib_loc\u001b[38;5;241m=\u001b[39mlib_loc)\n",
      "\u001b[1;31mPackageNotInstalledError\u001b[0m: The R package \"biomaRt\" is not installed."
     ]
    }
   ],
   "source": [
    "# Load the biomaRt library\n",
    "from rpy2.robjects.packages import importr\n",
    "biomaRt = importr('biomaRt')\n",
    "\n",
    "# Connect to the biomart database\n",
    "ensembl = biomaRt.useMart(\"ENSEMBL_MART_ENSEMBL\", dataset=\"hsapiens_gene_ensembl\")\n",
    "\n",
    "# Get ensemble genes from count.txt\n",
    "ensLookup = list(cts.rownames())\n",
    "\n",
    "# Get hgnc_symbol from Biomart\n",
    "annotLookup = biomaRt.getBM(attributes=[\"ensembl_gene_id\", \"gene_biotype\", \"hgnc_symbol\"],\n",
    "                            filters=\"ensembl_gene_id\", \n",
    "                            values=ensLookup, \n",
    "                            mart=ensembl, \n",
    "                            unique_rows=True)\n",
    "\n",
    "# Print the number of genes annotated by Biomart\n",
    "print(nrow(annotLookup), \"genes have been annotated by Biomart\")\n",
    "\n",
    "# Get counts.txt genes not been annotated by Biomart\n",
    "non_annotated_genes = list(set(cts.rownames()) - set(annotLookup[:,0]))\n",
    "print(len(non_annotated_genes), \"counts.txt Genes not annotated by Biomart\")\n",
    "\n",
    "# Add 'non_annotated_genes' as empty row to the annotLookup by Biomart\n",
    "empty_anno = pd.DataFrame({'ensembl_gene_id':non_annotated_genes,\n",
    "                           'gene_biotype': [None] * len(non_annotated_genes),\n",
    "                           'hgnc_symbol': [None] * len(non_annotated_genes)})\n",
    "annotLookup_with_empty_rows = pd.concat([annotLookup, empty_anno], axis=0, ignore_index=True)\n",
    "\n",
    "# Biomart exported 3 extra duplicated rows (annotations) with different 'hgnc_symbol'\n",
    "print(\"Biomart output of gene annotation contained\", len(annotLookup_with_empty_rows), \"rows of 57992 genes\")\n",
    "\n",
    "# Remove duplicated rows (first remained)\n",
    "print(\"Removing duplicated rows (first remained)\")\n",
    "annotLookup_with_empty_rows.drop_duplicates(subset=\"ensembl_gene_id\", keep=\"first\", inplace=True)\n",
    "\n",
    "# Finally, 57992 ensemble genes from count.txt with distinct symbols\n",
    "print(\"finally, 57992 ensemble genes from count.txt with distinct symbols\")\n",
    "\n",
    "# Setting genes in order same as in count.txt\n",
    "annotLookup_with_empty_rows.set_index(\"ensembl_gene_id\", inplace=True)\n",
    "annotLookup_with_empty_rows = annotLookup_with_empty_rows.reindex(cts.rownames())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98459c55-7400-49af-974e-ca805cfef02a",
   "metadata": {},
   "source": [
    "Skip to content\n",
    "Search or jump to…\n",
    "Pull requests\n",
    "Issues\n",
    "Codespaces\n",
    "Marketplace\n",
    "Explore\n",
    " \n",
    "@Libby-K \n",
    "ArielOgran\n",
    "/\n",
    "Champ-exercise\n",
    "Public\n",
    "Fork your own copy of ArielOgran/Champ-exercise\n",
    "Code\n",
    "Issues\n",
    "Pull requests\n",
    "Actions\n",
    "Projects\n",
    "Security\n",
    "Insights\n",
    "Beta Try the new code view\n",
    "Champ-exercise/Champions Oncology - exercise for bioinformatics candidates.Rmd\n",
    "@ArielOgran\n",
    "ArielOgran Add files via upload\n",
    "Latest commit 9681ea6 on Mar 4, 2022\n",
    " History\n",
    " 1 contributor\n",
    "299 lines (251 sloc)  13 KB\n",
    "\n",
    "---\n",
    "title: \"Chanpions Oncology - Exercise for bioinformatics candidates\"\n",
    "author: \"Arik Ogran\"\n",
    "date: \"3/1/2022\"\n",
    "output:\n",
    "  html_document: default\n",
    "  pdf_document: default\n",
    "---\n",
    "\n",
    "```{r setup, include=FALSE}\n",
    "knitr::opts_chunk$set(echo = TRUE)\n",
    "```\n",
    "\n",
    "Useful shortcut\n",
    "```{r}\n",
    "`%notin%` <- Negate(`%in%`)\n",
    "```\n",
    "\n",
    "### Installing R base packages which are not yet installed\n",
    "```{r message = FALSE, warning = FALSE}\n",
    "# R Package to load\n",
    "R_packages <- c(\"ggplot2\", \"stats\", \"dplyr\", \"tidyr\", \"ggfortify\", \"knitr\", \"lubridate\", \"imputeMissings\", \"stringr\", \"assist\", \"ggstatsplot\", \"forcats\", \"remedy\",  \"magrittr\", \"tidyverse\", \"devtools\", \"kableExtra\", \"pheatmap\")\n",
    "# Install packages not yet installed\n",
    "R_packages_to_install <- R_packages %notin% rownames(installed.packages())\n",
    "if (any(R_packages_to_install == TRUE)) {\n",
    "  install.packages(R_packages[R_packages_to_install])\n",
    "}\n",
    "# R Packages loading\n",
    "invisible(lapply(R_packages, library, character.only = TRUE))\n",
    "```\n",
    "\n",
    "### Installing Bioconductor packages which are not yet installed\n",
    "```{r warning = FALSE, message = FALSE}\n",
    "# # Bioconductor installation\n",
    "# if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "#   install.packages(\"BiocManager\")\n",
    "# BiocManager::install(version = \"3.14\")\n",
    "# BioC Package to load\n",
    "BioC_packages <- c(\"DESeq2\", \"edgeR\",\"biomaRt\", \"EnhancedVolcano\")\n",
    "# Install packages not yet installed\n",
    "BioC_already_exists <- installed.packages() %>% rownames() %>% devtools::package_info() %>% filter(source == \"Bioconductor\") %>% as.data.frame() %>% rownames()\n",
    "BioC_packages_to_install <- BioC_packages %notin% BioC_already_exists\n",
    "if (any(BioC_packages_to_install == TRUE)) {\n",
    "  BiocManager::install(BioC_packages[BioC_packages_to_install])\n",
    "}\n",
    "# BioC Packages loading\n",
    "invisible(lapply(BioC_packages, library, character.only = TRUE))\n",
    "```\n",
    "\n",
    "### Load the data into R and make sure the count and annotation data are consistentwith each other.\n",
    "```{r}\n",
    "# Setting working directory to hold raw data and outputs\n",
    "setwd(\"~/iCloud/Bioinformatics courses and tools/Chanpions Oncology - Exercise for bioinformatics candidates\")\n",
    "counts_df <- read.delim(\"counts.txt\")\n",
    "# converting counts_df to matrix\n",
    "rownames(counts_df) <- counts_df[,1]\n",
    "cts <- as.matrix(counts_df[,-1])\n",
    "coldata <- read.delim(\"sample-annotation.txt\")\n",
    "rownames(coldata) <- coldata[,1]\n",
    "gene_anno <- read.delim(\"gene-annotation.txt\")\n",
    "rownames(gene_anno) <- gene_anno[,1]\n",
    "message(\"Files have been loaded- count.txt, sample-annotation.txt and gene-annotation.txt\")\n",
    "```\n",
    "\n",
    "#### Missing data: gene-annotation.txt file is missing allot of genes which are presented in the counts.txt file\n",
    "```{r}\n",
    "cat(\"total ensemble genes in gene-annotation.txt:\", nrow(cts)) #57992 genes\n",
    "cat(\"total non-duplicated ensemble genes in gene-annotation.txt:\", rownames(cts) %>% unique() %>% length()) #57992 \n",
    "cat(\"Total genes annotated in gene-annotation.txt file:\", nrow(gene_anno)) #25503 genes\n",
    "cat(\"Genes from counts.txt file that does not have annotation in gene-annotation.txt:\", sum(rownames(cts) %notin% rownames(gene_anno))) # 32489 genes\n",
    "```\n",
    "\n",
    "#### Fetching more gene symbols using Biomart.\n",
    "Reason: In ExpressionSet, the number of rows in featureData must match the number of rows in assayData\n",
    "```{r}\n",
    "# Biomart\n",
    "mart <- useMart(\"ENSEMBL_MART_ENSEMBL\")\n",
    "mart <- useDataset(\"hsapiens_gene_ensembl\", mart)\n",
    "# ensemble genes (from count.txt) to annotate by Biomart\n",
    "ensLookup <- rownames(cts)\n",
    "# Getting hgnc_symbol from Biomart\n",
    "annotLookup <- getBM(mart=mart, attributes=c(\"ensembl_gene_id\",\"gene_biotype\",\"hgnc_symbol\"), # \"external_gene_name\"\n",
    "  filter=\"ensembl_gene_id\",\n",
    "  values=ensLookup,\n",
    "  uniqueRows=TRUE)\n",
    "cat(nrow(annotLookup), \"genes have been annotated by Biomart\") # 57051\n",
    "# counts.txt Genes not been annotated by Biomart\n",
    "non_annotated_genes <- rownames(cts)[which(rownames(cts) %notin% annotLookup[,1])] # 944 genes\n",
    "cat(length(non_annotated_genes), \"counts.txt Genes not annotated by Biomart\" )\n",
    "# adding 'non_annotated_genes' as empty row to the annotLookup by Biomart\n",
    "empty_anno <- data.frame(ensembl_gene_id=non_annotated_genes, gene_biotype=rep(NA, length(non_annotated_genes)), hgnc_symbol=rep(NA, length(non_annotated_genes)))\n",
    "# adding non annotated ensemble gene to the annotLookup\n",
    "annotLookup_with_empty_rows <- rbind(annotLookup, empty_anno)\n",
    "# Biomart exported 3 extra duplicated rows (annotations) with different 'hgnc_symbol'\n",
    "cat(\"Biomart output of gene annotation contained\", nrow(annotLookup_with_empty_rows), \"rows of 57992 genes\")\n",
    "# remove duplicated rows (first remained)\n",
    "message(\"Removeing duplicated rows (first remained)\")\n",
    "annotLookup_with_empty_rows <- annotLookup_with_empty_rows %>% distinct(ensembl_gene_id, .keep_all = TRUE)\n",
    "nrow(annotLookup_with_empty_rows) #finally, 57992 ensemble genes with distinct symbols\n",
    "cat(\"finally, 57992 ensemble genes from count.txt with distinct symbols\")\n",
    "#setting genes in order same as in count.txt\n",
    "rownames(annotLookup_with_empty_rows) <- annotLookup_with_empty_rows$ensembl_gene_id\n",
    "annotLookup_with_empty_rows <- annotLookup_with_empty_rows[rownames(cts),]\n",
    "```\n",
    "\n",
    "### Contruction of S4 object and quality control  \n",
    "#### Does count matrix and column data consistent in terms of sample order?\n",
    "It is absolutely critical that the columns of the count matrix and the rows of the column data (information about samples) are in the same order. \n",
    "```{r}\n",
    "# Test and set samples in order in count.txt and sample-annotation.txt\n",
    "if (any(!colnames(cts) == rownames(coldata))){\n",
    "  cts <- cts[,rownames(coldata)]\n",
    "}\n",
    "if (all(rownames(coldata) == colnames(cts))){\n",
    "  print(\"Sample order in count.txt and sample-annotation.txt is check and found consistant\")\n",
    "}\n",
    "```\n",
    "\n",
    "#### ExpressionSet object building and coercion to SummarizedExperiment\n",
    "```{r}\n",
    "# setting matadata for phenoData\n",
    "metadata <- data.frame(labelDescription=c(\"samples\",\"skin type\"),\n",
    "                       row.names=c(\"sample_id\", \"type\"))\n",
    "phenoData <- new(\"AnnotatedDataFrame\",data=coldata, varMetadata=metadata) # type: lesional vs normal\n",
    "# setting features for gene annotations\n",
    "features <- data.frame(labelDescription=c(\"ensembl_gene_id\", \"gene_biotype\",\"hgnc_symbol\"),\n",
    "                       row.names=c(\"ensembl_gene_id\", \"gene_biotype\",\"hgnc_symbol\"))\n",
    "annotation <- new(\"AnnotatedDataFrame\",data=annotLookup_with_empty_rows, varMetadata=features)\n",
    "# building object\n",
    "GSE54456_RNAseq <- ExpressionSet(assayData=cts,\n",
    "                                 phenoData=phenoData,\n",
    "                                featureData=annotation) \n",
    "# exprs(GSE54456_RNAseq)\n",
    "# phenoData(GSE54456_RNAseq)\n",
    "# sampleNames(GSE54456_RNAseq)\n",
    "# featureNames(GSE54456_RNAseq)\n",
    "# coercing to SummarizedExperiment\n",
    "GSE54456_RNAseq_SE <- makeSummarizedExperimentFromExpressionSet(GSE54456_RNAseq)\n",
    "message(\"ExpressionSet object has been built and coerced to SummarizedExperiment\")\n",
    "# setting sample type as factor\n",
    "GSE54456_RNAseq_SE$type <-  as.factor(GSE54456_RNAseq_SE$type)\n",
    "```\n",
    "\n",
    "#### Subset by support: Filter the count data for lowly-expressed genes, for example, only keep genes with a CPM >= 1 in at least 75% samples, in at least one of the groups.\n",
    "```{r}\n",
    "# computing CPM\n",
    "cpm <- apply(cts, 2, \n",
    "             function(x) x/sum(as.numeric(x)) * 10^6)\n",
    "# Check that the sum of each column (sample) after normalization equals to 10^6\n",
    "# colSums(cpm) \n",
    "# assign new \"cpm\" assay\n",
    "assay(GSE54456_RNAseq_SE, \"cpm\") <- cpm\n",
    "message(\"CPM has assigned to new assay of  GSE54456_RNAseq_SE SummarizedExperiment\")\n",
    "# samples count per skin type  \n",
    "lesional_samples <- sum(colData(GSE54456_RNAseq_SE)$type == \"lesional\")\n",
    "normal_samples <- sum(colData(GSE54456_RNAseq_SE)$type == \"normal\")\n",
    "# Calculate support for lesional group\n",
    "rowData(GSE54456_RNAseq_SE)$lesional_CPM_support <- as.integer(rowSums(assay(GSE54456_RNAseq_SE[,GSE54456_RNAseq_SE$type == \"lesional\"], \"cpm\") >= 1))\n",
    "# Calculate support for normal group\n",
    "rowData(GSE54456_RNAseq_SE)$normal_CPM_support <- as.integer(rowSums(assay(GSE54456_RNAseq_SE[,GSE54456_RNAseq_SE$type == \"normal\"], \"cpm\") >= 1))\n",
    "# gene count before filtering out lowly expressed genes\n",
    "before <- nrow(GSE54456_RNAseq_SE)\n",
    "    \n",
    "# Subset\n",
    "GSE54456_RNAseq_SE_filtered <- GSE54456_RNAseq_SE[which(100 * (rowData(GSE54456_RNAseq_SE)$lesional_CPM_support / lesional_samples) >= 75 | \n",
    "                                                  100 * (rowData(GSE54456_RNAseq_SE)$normal_CPM_support / normal_samples) >= 75)  , ]\n",
    "message(\"Lowly-expresssed genes have been filtered out from GSE54456_RNAseq_SE \")\n",
    "after <- nrow(GSE54456_RNAseq_SE_filtered)\n",
    "removed <- before - after\n",
    "    \n",
    "# Print summary for subset by support \n",
    "message(\"Removed \", removed, \" out of \", before, \" genes (\", round(removed/before * 100, digits = 1), \"%)\")\n",
    "```\n",
    "\n",
    "### Generate an object that contains the library-size normalized log-CPM data. Saveit as a binary file (.rda or .rds).\n",
    "```{r}\n",
    "log_cpm <- apply(cts, 2, \n",
    "             function(x) log(x/sum(as.numeric(x)) * 10^6))\n",
    "save(log_cpm, file=\"logCPM.rda\")\n",
    "message(\"Binnary file of library-size normalized log-CPM data has been exported to working directory\")\n",
    "```\n",
    "\n",
    "### The PCA plot may suggest the presence of outlier/mis-labeled samples in this dataset. Try to identify them and remove them from the downstream analysis.\n",
    "Before performing statistical tests for various measures of Differential Expression (DE), it is important to first conduct a thorough Exploratory Data Analysis (EDA) to identify what factors we need to include in the final DE model. In this exercise case we wont look for a factor but only mis-labaled samples to remove.\n",
    "```{r}\n",
    "# transpose the matrix \n",
    "M <- t(cpm)\n",
    "# transform the counts to log2 scale \n",
    "M <- log2(M + 1)\n",
    "# compute PCA \n",
    "pcaResults <- prcomp(M)\n",
    "# plot PCA results making use of ggplot2's autoplot function\n",
    "# ggfortify is needed to let ggplot2 know about PCA data structure. \n",
    "autoplot(pcaResults, data = coldata, colour = 'type')\n",
    "cat(\"Except one sample, we can be split the samples into two distinct clusters on the scatter plot based on largest principal components 1\")\n",
    "# identify the outlier/mis-labeled sample\n",
    "mis_labeled <- rownames(pcaResults$x)[pcaResults$x[,\"PC1\"] < -0.05 & rownames(pcaResults$x) %in% coldata$sample_id[coldata$type == \"lesional\"]]\n",
    "# removing outlier/mis-labeled sample/s\n",
    "GSE54456_RNAseq_SE_filtered <- GSE54456_RNAseq_SE_filtered[,which(GSE54456_RNAseq_SE_filtered$sample_id != mis_labeled)]\n",
    "message(\"outlier/mis-labeled sample has been removed\")\n",
    "```\n",
    "\n",
    "### Differential expression analysis\n",
    "Export the results in a tab-separated text/CSV file: a table with genes in rowsalongwith gene annotations and any relevant statistic.\n",
    "```{r message = FALSE, warning = FALSE}\n",
    "# Building DESeq object from SummarizedExperiment\n",
    "ddsSE <- DESeqDataSet(GSE54456_RNAseq_SE_filtered, design = ~ type)\n",
    "# Normalize and log transform\n",
    "vst_blind <- vst(ddsSE, blind = TRUE)\n",
    "# PCA-plot of variance stabilized expression\n",
    "plotPCA(vst_blind, \"type\")\n",
    "cat(\"No batch effect was found to be included in the DESeq design\")\n",
    "# compute the contrast for the 'type' variable where 'normal' samples are used as the control group.\n",
    "# results are binded to rowData to get gene symbols\n",
    "dds <- DESeq(ddsSE)\n",
    "DEresults = results(dds, contrast = c(\"type\", 'lesional', 'normal'),  tidy = TRUE) %>%\n",
    "                     bind_cols(as.data.frame(rowData(GSE54456_RNAseq_SE_filtered))) %>%\n",
    "                     as_tibble()\n",
    "#sort results by increasing p-value\n",
    "DEresults <- DEresults[order(DEresults$pvalue),]\n",
    "# How many adjusted p-values were less than 0.1?\n",
    "cat(sum(DEresults$padj < 0.1, na.rm=TRUE), \" genes are DE with < 0.1 adjusted p-values\")\n",
    "# Exporting to results to .csv file\n",
    "write.csv(DEresults,\"DEresults.csv\")\n",
    "```\n",
    "\n",
    "### Select the top 100 most significant annotated genes and generate a heatmap of the log-CPM data, with samples in columns, annotated with the group variable.\n",
    "```{r}\n",
    "# Show top 100 \n",
    "DEresults %>%\n",
    "  top_n(n = -100, wt = padj) %>%\n",
    "  dplyr::select(Gene = row, \n",
    "                Symbol = hgnc_symbol,\n",
    "                baseMean, \n",
    "                log2FoldChange, \n",
    "                padj) %>%\n",
    "  kable(caption = \"Top differentially expressed genes\") %>%\n",
    "  kable_styling(latex_options = \"hold_position\")\n",
    "# Top 100\n",
    "Top_100 <- DEresults %>% arrange (desc(abs(log2FoldChange))) %>% head (100)\n",
    "# log_cpm pheatmap:\n",
    "# Since Count data was filtered to hold genes that expressed in >= 1 CPM in at list 75% of samples, some genes have zero count in some samples which result in log_CPM as -INF.\n",
    "# In order to pheatmap log_cpm, -Inf values must be changed to 0\n",
    "noInf <- log_cpm\n",
    "noInf[which(!is.finite(noInf))] <- 0\n",
    "# setting annotation_col\n",
    "df <- as.data.frame(colData(ddsSE)[,c(\"sample_id\",\"type\")])\n",
    "# ploting pheatmap\n",
    "pheatmap(noInf[Top_100$row,], cluster_rows=TRUE, show_rownames=FALSE, show_colnames=FALSE,\n",
    "         cluster_cols=T, annotation_col=df)\n",
    "```\n",
    "\n",
    "### Generate a volcano plot (x-axis is the effect size and y-axis is the p-value) for this analysis. The selected 100 most significant genes should be colored.\n",
    "```{r}\n",
    "# setting red color  and size for top 100 DE genes\n",
    "cols <- ifelse(DEresults$row %in% Top_100$row, 'red', 'black')\n",
    "size <- ifelse(DEresults$row %in% Top_100$row, 1.2, 0.6)\n",
    "# plotting volcano\n",
    "plot(DEresults$log2FoldChange, -log10(DEresults$padj), col=cols, panel.first=grid(),\n",
    "     main=\"Volcano plot\", xlab=\"Effect size: log2(fold-change)\", ylab=\"-log10(adjusted p-value)\",\n",
    "     pch=20, cex=size)\n",
    "abline(v=0, col=\"black\", lty=3, lwd=1.0)\n",
    "abline(v=3.5, col=\"black\", lty=4, lwd=2.0)\n",
    "abline(v=-3.5, col=\"black\", lty=4, lwd=2.0)\n",
    "```\n",
    "\n",
    "\n",
    "```{r}\n",
    "sessionInfo()\n",
    "```\n",
    "\n",
    "Footer\n",
    "© 2023 GitHub, Inc.\n",
    "Footer navigation\n",
    "Terms\n",
    "Privacy\n",
    "Security\n",
    "Status\n",
    "Docs\n",
    "Contact GitHub\n",
    "Pricing\n",
    "API\n",
    "Training\n",
    "Blog\n",
    "About\n",
    "Champ-exercise/Champions Oncology - exercise for bioinformatics candidates.Rmd at main · ArielOgran/Champ-exercise"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
